### ISSUE TITLE
Build-Push-Action / Multiarch

### ISSUE THREAD

**Issue by soywod** [tomav/docker-mailserver#1092]
When starting the server, got `standard_init_linux.go:190: exec user process caused "exec format error"`.

## Context

Install docker and docker-compose on raspberry (OS raspbian), install and start the server

## Expected Behavior

Should start

## Actual Behavior

Not starting

## Your Environment

* Amount of RAM available: 2GB
* Mailserver version used: master
* Docker version used: 18.09.0
* Environment settings relevant to the config: Rasberry (arm) on raspbian
* Any relevant stack traces ("Full trace" preferred):

```
standard_init_linux.go:190: exec user process caused "exec format error"
```

**Comment by johansmitsnl** [comment]
There is on #348 issue (closed). I personally have no arm devices so I don't know if it works. I assume you need to build your image yourself. On the docker hub it is only x86_64?

**Comment by soywod** [comment]
Yes it's what I found yesterday, browsing your issues. I started to do so, but seems to be quite complicated (especially for filebeat). I will dig a bit. You think it could be possible to add a `arm` version of your `Dockerfile` in the repo?

**Comment by jnovack** [comment]
See [mariadb-docker-armhf](https://github.com/davidecavestro/mariadb-docker-armhf/tree/756c3f31a117341896c37976640055fb43d5b004) for inspiration.

```
.
├── Dockerfile
└── hooks
     ├── post_checkout
     └── pre_build
```
`hooks/post_checkout`
```
#!/bin/bash
# downloads a local copy of qemu on docker-hub build machines
curl -L https://github.com/balena-io/qemu/releases/download/v3.0.0%2Bresin/qemu-3.0.0+resin-arm.tar.gz | tar zxvf - -C . && mv qemu-3.0.0+resin-arm/qemu-arm-static .
```
`hooks/pre_build`
```
#!/bin/bash
# Register qemu-*-static for all supported processors except the 
# current one, but also remove all registered binfmt_misc before
docker run --rm --privileged multiarch/qemu-user-static:register --reset
```
`Dockerfile`
```
FROM arm32v7/debian:stretch-slim 
COPY qemu-arm-static /usr/bin
...
```

**Comment by maxswjeon** [comment]
Is this going to be merged to main branch or not? Should I make those changes by myself?

**Comment by edgeux** [comment]
couple of things... i also got this to eventually build by tweaking this Dockerfile slightly. 

1) i had to build filebeat first and then COPY that into the final image. 

#----- Create a Docker for cross-compilation -----#
mkdir build && cd $_
docker run -it --rm -v `pwd`:/build golang:1.8.3 /bin/bash
#----- Inside docker -----#
go get github.com/elastic/beats       
cd /go/src/github.com/elastic/beats/filebeat/
git checkout v5.6.3
GOARCH=arm go build
cp filebeat /build
exit

2) i had could not get that base image to play so i used straight DEBIAN instead. this installed all the packages without a hitch:

FROM debian

3) lastly you have do a full clone: 

git clone --recurse-submodules https://github.com/tomav/docker-mailserver.git

i will be testing in due course.

**Comment by unomind** [comment]
any news?

**Comment by edgeux** [comment]
update: ive been running the mailserver on a rpi4 now for few weeks without a single glitch. totally hassle free. The footprint on the hardware is minimal.

i initially tried on rpi3 x 3  docker swarm and it struggled with the shared storage and the permissions were just too much for my nooby mind. 

but rpi4 is perfect for a small docker server.

**Comment by unomind** [comment]
> update: ive been running the mailserver on a rpi4 now for few weeks without a single glitch. totally hassle free. The footprint on the hardware is minimal.
> 
> i initially tried on rpi3 x 3 docker swarm and it struggled with the shared storage and the permissions were just too much for my nooby mind.
> 
> but rpi4 is perfect for a small docker server.

Whould you mind to share your workaround? Maybe you have blog etc. than I can read.

Thanks

**Comment by edgeux** [comment]
The steps are right there in my previous post. 
1. build filebeat first. you can do this with a docker cross compiler. 
2. i tweaked the docker file from @soywod  to just 'debian' base image and swap the filebeat part to simply COPY into place your freshly built filebeat binary.
3. On the PI, build the image ( dont build it on anything else or youll just make a headache for yourself. )
4. done, just fire it up without the -d flag first to check for debug and errors. all other official guides and howtos apply.

i would suggest not using clamav to start with if youre not on rpi4. itll just slow things down. 

once its running pull it down and then use the -d

**Comment by unomind** [comment]
Cool, thank you. Step by step tutorial will be good for new comer.

**Comment by EliSu** [comment]
Really cool all you guys~

**Comment by erik-wramner** [comment]
Question if this is going into the main branch - can Travis build and test the ARM version? How do we know that new builds actually work? Personally I feel it would work best as a fork that merges changes from upstream with manual tests (or some CI engine that runs on ARM) before pushing. Alternative we could keep it here as a separate branch with the same kind of manual merge procedure. Automatically pushing all changes directly to the ARM image (i.e. building both from the same source) without reasonably high test coverage on both platforms feels dangerous.

Cool that it works though!

**Comment by fbartels** [comment]
Travis could build for arm, but I have not yet played with that myself: https://dev.to/zeerorg/build-multi-arch-docker-images-on-travis-5428

**Comment by VincentCATILLON** [comment]
FYI, you can use `docker buildx` for example to build the docker image for some platforms:

```
git clone git@github.com:tomav/docker-mailserver.git
cd docker-mailserver
docker buildx build --platform linux/arm/v7,linux/arm64/v8 -f Dockerfile -t <your_scope>/docker-mailserver --push .
```

And so, check that everything works:

```
docker run <your_scope>/docker-mailserver /bin/sh 
```

**Comment by radicand** [comment]
For those that want something that "just works" on ARM, I put together learnings from this thread here: https://github.com/radicand/docker-mailserver and published an image. You can use `radicand/docker-mailserver` for the image name instead, but all the rest of the configuration should work as is. I'm running it in a K8s (k3s) cluster at the moment without issue.

**Comment by ktcar214** [comment]
Where I should copy compiled filebeat? I'm new to docker and don't know where and how I should do this.

**Comment by edgeux** [comment]
> Where I should copy compiled filebeat? I'm new to docker and don't know where and how I should do this.

You need to read up basics on docker files. Building the the image will take the filebeat files from your host filesystem and move it into place in the docker image. this is done with the COPY function.

**Comment by erik-wramner** [comment]
@ktcar214 note that filebeat will probably be moved out of the container in a coming release. You might as well run filebeat in another docker image. With that approach you can use a pre-built filebeat image. The official one does not support ARM, but there are bound to be others. See #1321.

**Comment by bertovanoorspronk** [comment]
Removing filebeat for running on raspberry is a good idea. Filebeat its a big problem if you want to run the original Docker file.

If i upgrade from stretch to buster  (FROM debian:buster-slim) in the Docker file i get a problem with the software packages zoo and ripole. When i remove this packages, i can run the mail-server on Raspberry pi again.

Are the packages zoo and ripole required?

**Comment by radicand** [comment]
> If i upgrade from stretch to buster (FROM debian:buster-slim)

I attempted this in my fork, but the dovecot version was an issue (possibly other versioning issues too, I can't recall) and the CE apt repo does not have arm support - something to check if you believe it works out of the box. I ended up staying with stretch and using the dovecot from stretch-backports.

My Dockerfile for reference: https://github.com/radicand/docker-mailserver/blob/master/Dockerfile

**Comment by erik-wramner** [comment]
@bertovanoorspronk I'm also trying to upgrade to buster and ran into zoo and ripole. They are not really required, they are only used by amavis for analyzing e-mails. Nobody uses zoo archives anymore, but ripole is used to inspect Word documents. That would have been nice to have, but I can't find a replacement.

In short you can build the image without them, it works but it may be harder to detect viruses hidden in Word documents.

**Comment by radicand** [comment]
To update this thread, the new master based on Buster builds fine as-is now for arm64. I have been doing a custom ARM build for awhile, but now am building directly against master. I have issues building for ARMv7, but given that any modern RPi can enable a 64 bit kernel now, I'm satisfied with just building ARM64 only.

https://hub.docker.com/repository/docker/radicand/docker-mailserver/tags - the `gha` tag is the new mirrored build.

**Comment by nalakawula** [comment]
RPi 3 also capable to run 64 bit os by using Ubuntu server image, [https://ubuntu.com/download/raspberry-pi](https://ubuntu.com/download/raspberry-pi/)

**Comment by edgeux** [comment]
on RPI4 ive tried a fresh ubuntu(arm64)/docker/compose and build always fails at step 49/56:

Step 49/56 : COPY ./target/helper_functions.sh ./target/check-for-changes.sh ./target/start-mailserver.sh ./target/fail2ban-wrapper.sh ./target/postfix-wrapper.sh ./target/postsrsd-wrapper.sh ./target/docker-configomat/configomat.sh /usr/local/bin/
COPY failed: stat /var/lib/docker/tmp/docker-builder254095452/target/docker-configomat/configomat.sh: no such file or directory

any thoughts from anyone?

**Comment by radicand** [comment]
> if youre building for arm64 and you get something like this at step 49
> 
> Step 49/56 : COPY ./target/helper_functions.sh ./target/check-for-changes.sh ./target/start-mailserver.sh ./target/fail2ban-wrapper.sh ./target/postfix-wrapper.sh ./target/postsrsd-wrapper.sh ./target/docker-configomat/configomat.sh /usr/local/bin/
> COPY failed: stat /var/lib/docker/tmp/docker-builder254095452/target/docker-configomat/configomat.sh: no such file or directory
> 
> you need to make sure the /target/docker-configomat/configomat.sh script is present by doing a full clone and including linked files.
> 
> you can manually add this file to the target path and rebuild and it should work.

Ensuring you clone the submodules will resolve this.

Separately though, I think the conversation in this thread should shift to multi arch builds in the main build pipeline now, given it works with no changes.

**Comment by erik-wramner** [comment]
Indeed, good work!

**Comment by fergalmoran** [comment]
@radicand - thanks so much for the arm64 builds, works like magic...
Off topic but I notice you said you had this working in k3s? Any chance you could share your config?

**Comment by radicand** [comment]
@fergalmoran - here's my all-in-one deployment file: https://gist.github.com/radicand/9f5ccd4650479fe9a0a4b762ad1e6be9 - i've removed sensitive data and you'll need to replace values appropriately. HTH!

**Comment by fergalmoran** [comment]
Thank you so much @radicand - HUGELY appreciated!

**Comment by radicand** [comment]
@erik-wramner , do you know if your build pipeline supports Docker buildx? If it's Docker 19.03 or higher, it should. If that's the case, if you can also confirm the build command is issued from the Makefile is what's used in your build/deploy pipeline, I can submit a PR to update the syntax to support arm64.

**Comment by erik-wramner** [comment]
Travis seems to use Docker 18.06 on Ubuntu 16.04. I'm not 100% sure how the builds work. The tests run from the Makefile, but there could be another step that actually publishes to docker hub. There is nothing in the build logs about the publishing step. @fbartels or @tomav do you know?

**Comment by fbartels** [comment]
The final image is built directly on the docker hub at https://github.com/tomav/docker-mailserver/blob/master/hooks/build. So buildx should be supported, afaik the hook needs to be extended for multiple architectures.

Its been  quite a while since I had a closer look at the dockerfile, but i imagine that some of it would need to be updated for it as well.

**Comment by radicand** [comment]
@fbartels if there's a way to test this, it should work out of the box - I have been building without changes since the buster build using this buildx command: https://github.com/radicand/docker-mailserver/blob/master/.github/workflows/dockerimage.yml#L35-L41

**Comment by fbartels** [comment]
ok, then is should be a matter of updating the build hook with the lines you have in your action. I don't think this can be tested however, since afair the hook is only triggered when something is pushed into the master branch.

**Comment by radicand** [comment]
@fbartels - I'll take it as an action to modify the hook on my fork and report back.

**Comment by radicand** [comment]
Unfortunately, DockerHub is on 18.03 (https://github.com/docker/hub-feedback/issues/1874#issuecomment-632602168) and does not support buildx, my suggestion is to switch from DockerHub-based builds and go with a GitHub action instead (see my repo link above) which will handle the build+push properly.

**Comment by erik-wramner** [comment]
Is it possible to schedule regular builds (not triggered by a commit) with that solution as well? That has been on my wish list for quite some time.

**Comment by radicand** [comment]
> Is it possible to schedule regular builds (not triggered by a commit) with that solution as well? That has been on my wish list for quite some time.

Yes, see here: https://github.com/radicand/docker-mailserver/blob/master/.github/workflows/sync-upstream.yml#L3-L5

**Comment by Yajo** [comment]
Yes, it's totally possible.

@radicand's link is broken, so let me share mine: https://github.com/Tecnativa/doodba/blob/ab9c74d1650b27c4d8bfb26d14d203be1844547f/.github/workflows/ci.yaml#L64-L83

It has nothing to do with mail, but you see how secrets and images can be used to push both to docker hub and github docker registry.

You have a full ubuntu VM to do the build, so you can do whatever you want.

Some interesting details also:

- There's [an official build-and-push action](https://github.com/docker/build-push-action).
- Still without buildx support: https://github.com/docker/build-push-action/issues/36 so probably it's better to do it "by hand" for now.
- Providing images in github registry has the additional benefit that people living in countries blocked by USA can still download your images.

**Comment by radicand** [comment]
Thanks for noting the broken link @Yajo - here's a link to the tree where cron and buildx are both present: https://github.com/radicand/docker-mailserver/tree/63a9ea44e3435ddfc68e1aaa7f98b41478f03d10/.github/workflows

**Comment by radicand** [comment]
@bertovanoorspronk it should be, I see it in my image. Are you using the :gha tag?

**Comment by aendeavor** [comment]
@radicand Great work! I was curious: Is the current conversation still related to the issue-topic, seeing that ARM builds work.

I would like to suggest opening a new issue about **buildx and GitHub automated builds**, as the current topic seems important.

**Comment by Yajo** [comment]
This keeps getting easier :blush: 

See https://github.com/docker/build-push-action/issues/36#issuecomment-685732033

**Comment by aendeavor** [comment]
Has the current project stalled or is there work in progress here?

**Comment by aendeavor** [comment]
This issue was closed due to one or more of the following reasons:

1. Age
2. Contributor inactivity
3. The issue seems to be resolved

If you think this happened by accident, or feel like this issue was not actually resolved, please feel free to re-open it. If there is an issue you could resolve in the meantime, please open a PR based on the current master branch so we can review it.
